RecoEgamma/EgammaElectronAlgos/src/ConversionFinder.cc

```diff
@@ -0,0 +1,300 @@
+  }
+
+  //------------------------------------------------------------------------------------
+
+  //takes in a vector of candidate conversion partners
+  //and arbitrates between them returning the one with the
+  //smallest R=sqrt(dist*dist + dcot*dcot)
+  ConversionInfo const& arbitrateConversionPartnersbyR(const std::vector<ConversionInfo>& convCandidates) {
+    ConversionInfo const* closestConversion = &convCandidates.front();
+
+    if (convCandidates.size() == 1)
+      return *closestConversion;
+
+    float R = pow(closestConversion->dist, 2) + pow(closestConversion->dcot, 2);
+
+    for (auto const& temp : convCandidates) {
+      float temp_R = pow(temp.dist, 2) + pow(temp.dcot, 2);
+      if (temp_R < R) {
+        R = temp_R;
+        closestConversion = &temp;
+      }
+    }
+
+    return *closestConversion;
+  }
+
+  //------------------------------------------------------------------------------------
+  ConversionInfo findBestConversionMatch(const std::vector<ConversionInfo>& v_convCandidates) {
+    using namespace std;
+
+    if (v_convCandidates.empty())
+      return ConversionInfo{-9999., -9999., -9999., std::nullopt, std::nullopt, -9999, -9999};
+
+    if (v_convCandidates.size() == 1)
+      return v_convCandidates.at(0);
+
+    vector<ConversionInfo> v_0;
+    vector<ConversionInfo> v_1;
+    vector<ConversionInfo> v_2;
+    vector<ConversionInfo> v_3;
+    //loop over the candidates
+    for (unsigned int i = 1; i < v_convCandidates.size(); i++) {
+      ConversionInfo temp = v_convCandidates.at(i);
+
+      if (temp.flag == 0) {
+        bool isConv = false;
+        if (std::abs(temp.dist) < 0.02 && std::abs(temp.dcot) < 0.02 && temp.deltaMissingHits < 3 &&
+            temp.radiusOfConversion > -2)
+          isConv = true;
+        else if (pow(temp.dist, 2) + pow(temp.dcot, 2) < square(0.05f) && temp.deltaMissingHits < 2 &&
+                 temp.radiusOfConversion > -2)
+          isConv = true;
+
+        if (isConv)
+          v_0.push_back(temp);
+      }
+
+      if (temp.flag == 1) {
+        if (pow(temp.dist, 2) + pow(temp.dcot, 2) < square(0.05f) && temp.deltaMissingHits < 2 &&
+            temp.radiusOfConversion > -2)
+          v_1.push_back(temp);
+      }
+      if (temp.flag == 2) {
+        if (pow(temp.dist, 2) + pow(temp.dcot * temp.dcot, 2) < square(0.05f) && temp.deltaMissingHits < 2 &&
+            temp.radiusOfConversion > -2)
+          v_2.push_back(temp);
+      }
+      if (temp.flag == 3) {
+        if (temp.dist * temp.dist + temp.dcot * temp.dcot < square(0.05f) && temp.deltaMissingHits < 2 &&
+            temp.radiusOfConversion > -2)
+          v_3.push_back(temp);
+      }
+
+    }  //candidate conversion loop
+
+    //now do some arbitration
+
+    //give preference to conversion partners found in the CTF collection
+    //using the electron's CTF track
+    if (!v_0.empty())
+      return arbitrateConversionPartnersbyR(v_0);
+
+    if (!v_1.empty())
```

Comment:
If v_0 is not empty, v_1, v_2, v_3 are not needed any more.
One can then gain some more time if, in the loop above:
- v_1 is only filled if _v_0 is still empty
- v_2 is only filled if both v_0 and v_1 are still empty
_ v_3 is only filled if v_0, v_1 and v_2 are still empty

Reply:
I'm not sure if we can really get some time here. Filling these `v_<i>` only when needed potentially requires multiple iterations over the full conversion list. I think profiling would be needed to decide if this optimization would work in our favor, but it would be difficult to measure any effect. The cost of the arbitration is negligible in igprof (0.00 %), the conversion finder cost is completely dominated by finding the full list of conversions (`getConversionInfos`):

https://rembserj.web.cern.ch/rembserj/cgi-bin/igprof-navigator/30956_23434.21_with_PR_igreport_perf/1672